### ISSUE TITLE
alpine linux get wrong errno but rbx log show right

### ISSUE THREAD

**Issue by SquallATF** [rubinius/rubinius#3735]
1. What command did you run?

  I ran `./staging/bin/rbx bin/mspec ci -t bin/rbx -V -d --background ./spec/ruby/core/kernel/spawn_spec.rb` and `./staging/bin/rbx bin/mspec ci -t bin/rbx -V -d --background /home/squallatf/rubinius-3.70/spec/ruby/core/process/spawn_spec.rb`

2. What behavior did you expect?

  I expected pass test

3. What behavior did you get instead?
```
rubinius 3.70 (2.3.1 58f85ba2 2017-01-24 3.8.1) [x86_64-linux-gnu]

/home/squallatf/rubinius-3.70/spec/ruby/core/kernel/spawn_spec.rb..........................................FFF..........................................................................FFF.................................

1)
Kernel#spawn raises an Errno::ENOENT if the command does not exist FAILED
Expected Errno::ENOENT
but got Errno::EINVAL (Invalid argument - execvp(2) failed)
                main # Rubinius::Loader at core/loader.rb:861
              script # Rubinius::Loader at core/loader.rb:679
         load_script . Rubinius::CodeLoader at core/code_loader.rb:590
         load_script # Rubinius::CodeLoader at core/code_loader.rb:505
          __script__ # Object at mspec/bin/mspec-ci:8
                main . MSpecScript at mspec/lib/mspec/utils/script.rb:217
                 run # MSpecCI at mspec/lib/mspec/commands/mspec-ci.rb:78
             process . MSpec at mspec/lib/mspec/runner/mspec.rb:43
               files . MSpec at mspec/lib/mspec/runner/mspec.rb:51
                each # Array at core/array.rb:72
        { } in files . MSpec at mspec/lib/mspec/runner/mspec.rb:57
             protect . MSpec at mspec/lib/mspec/runner/mspec.rb:69
       instance_eval # BasicObject(Object) at core/basic_object.rb:102
        { } in files . MSpec at mspec/lib/mspec/runner/mspec.rb:57
                load . Kernel at core/kernel.rb:572
          __script__ # Object at spec/ruby/core/kernel/spawn_spec.rb:5
            describe # Object at mspec/lib/mspec/runner/object.rb:11
            describe . MSpec at mspec/lib/mspec/runner/mspec.rb:38
             process # ContextState at mspec/lib/mspec/runner/context.rb:199
                each # Array at core/array.rb:72
      { } in process # ContextState at mspec/lib/mspec/runner/context.rb:200
              repeat . MSpec at mspec/lib/mspec/runner/mspec.rb:238
               times # Integer(Fixnum) at core/integer.rb:194
       { } in repeat . MSpec at mspec/lib/mspec/runner/mspec.rb:239
      { } in process # ContextState at mspec/lib/mspec/runner/context.rb:208
             protect # ContextState at mspec/lib/mspec/runner/context.rb:179
                all? # Enumerable(Array) at core/enumerable.rb:405
                each # Array at core/array.rb:72
         { } in all? # Enumerable(Array) at core/enumerable.rb:405
      { } in protect # ContextState at mspec/lib/mspec/runner/context.rb:179
             protect . MSpec at mspec/lib/mspec/runner/mspec.rb:69
       instance_eval # BasicObject(Object) at core/basic_object.rb:102
   { } in __script__ # Object at spec/ruby/shared/process/spawn.rb:533
              should # Object(Proc) at mspec/lib/mspec/expectations/should.rb:8
           fail_with . SpecExpectation at mspec/lib/mspec/expectations/expectations.rb:15

2)
Kernel#spawn raises an Errno::EACCES when the file does not have execute permissions FAILED
Expected Errno::EACCES
but got Errno::EINVAL (Invalid argument - execvp(2) failed)
...................................................

3)
Kernel#spawn raises an Errno::EACCES when passed a directory FAILED
Expected Errno::EACCES
but got Errno::EINVAL (Invalid argument - execvp(2) failed)
........................................

4)
Kernel.spawn raises an Errno::ENOENT if the command does not exist FAILED
Expected Errno::ENOENT
but got Errno::EINVAL (Invalid argument - execvp(2) failed)
........................................

5)
Kernel.spawn raises an Errno::EACCES when the file does not have execute permissions FAILED
Expected Errno::EACCES
but got Errno::EINVAL (Invalid argument - execvp(2) failed)
......................................

6)
Kernel.spawn raises an Errno::EACCES when passed a directory FAILED
Expected Errno::EACCES
but got Errno::EINVAL (Invalid argument - execvp(2) failed)
.....................................

Finished in 24.555472 seconds

1 file, 155 examples, 185 expectations, 6 failures, 0 errors
```
and
```
rubinius 3.70 (2.3.1 58f85ba2 2017-01-24 3.8.1) [x86_64-linux-gnu]

/home/squallatf/rubinius-3.70/spec/ruby/core/process/spawn_spec.rb.........................................FFF.................................

1)
Process.spawn raises an Errno::ENOENT if the command does not exist FAILED
Expected Errno::ENOENT
but got Errno::EINVAL (Invalid argument - execvp(2) failed)
......................

2)
Process.spawn raises an Errno::EACCES when the file does not have execute permissions FAILED
Expected Errno::EACCES
but got Errno::EINVAL (Invalid argument - execvp(2) failed)
......................................

3)
Process.spawn raises an Errno::EACCES when passed a directory FAILED
Expected Errno::EACCES
but got Errno::EINVAL (Invalid argument - execvp(2) failed)
........................................

Finished in 12.231985 seconds

1 file, 77 examples, 92 expectations, 3 failures, 0 errors
```

but system log show right error info
part of `kernel/spawn_spec.rb` log
```
.....................................
Feb  6 21:50:51 [32635] process: spawn: 539: nonesuch, ruby.main, /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb:533
Feb  6 21:50:51 [539] <Error> No such file or directory: spawn: exec failed: nonesuch
Feb  6 21:50:51 [32635] process: spawn: 540: /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb, ruby.main, /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb:537
Feb  6 21:50:51 [540] <Error> Permission denied: spawn: exec failed: /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb
Feb  6 21:50:51 [32635] process: spawn: 541: /home/squallatf/rubinius-3.70/spec/ruby/shared/process, ruby.main, /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb:541
Feb  6 21:50:51 [541] <Error> Permission denied: spawn: exec failed: /home/squallatf/rubinius-3.70/spec/ruby/shared/process
..................................
Feb  6 21:51:04 [32635] process: spawn: 801: nonesuch, ruby.main, /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb:533
Feb  6 21:51:04 [801] <Error> No such file or directory: spawn: exec failed: nonesuch
Feb  6 21:51:04 [32635] process: spawn: 802: /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb, ruby.main, /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb:537
Feb  6 21:51:04 [802] <Error> Permission denied: spawn: exec failed: /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb
Feb  6 21:51:04 [32635] process: spawn: 803: /home/squallatf/rubinius-3.70/spec/ruby/shared/process, ruby.main, /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb:541
Feb  6 21:51:04 [803] <Error> Permission denied: spawn: exec failed: /home/squallatf/rubinius-3.70/spec/ruby/shared/process
....................................
```
and
part of `process/spawn_spec.rb` log
```
Feb  6 22:09:38 [1024] process: spawn: 1195: nonesuch, ruby.main, /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb:533
Feb  6 22:09:38 [1195] <Error> No such file or directory: spawn: exec failed: nonesuch
Feb  6 22:09:38 [1024] process: spawn: 1196: /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb, ruby.main, /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb:537
Feb  6 22:09:38 [1196] <Error> Permission denied: spawn: exec failed: /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb
Feb  6 22:09:38 [1024] process: spawn: 1197: /home/squallatf/rubinius-3.70/spec/ruby/shared/process, ruby.main, /home/squallatf/rubinius-3.70/spec/ruby/shared/process/spawn.rb:541
Feb  6 22:09:38 [1197] <Error> Permission denied: spawn: exec failed: /home/squallatf/rubinius-3.70/spec/ruby/shared/process
```

4. What version of Rubinius?
  
  `rubinius 3.70 (2.3.1 58f85ba2 2017-01-24 3.8.1) [x86_64-linux-gnu]`

5. What version of operating system?
  
  `Linux magi10 4.4.45-0-virtgrsec #1-Alpine SMP Thu Jan 26 14:32:43 GMT 2017 x86_64 GNU/Linux`

6. What is your operating system distribution, if your operating system has more than one?

7. How did you build your version of Rubinius?

  - [ ] I used RVM.
  - [ ] I used ruby-build.
  - [ ] I used ruby-install.
  - [x] I built manually from a tarball.
  - [ ] I built manually from a git clone.
  - [ ] I installed a binary.

8. Does this issue involve proprietary code?

  - [ ] Yes, this issue involves proprietary code that I cannot share.
  - [ ] Yes, this issue involves proprietary code, but I am able to share it under certain conditions.
  - [x] No, this issued doesn't involve proprietary code.

9. Are you able to help us debug the issue?

  - [ ] Yes, I'm able to help debug, including running commands under lldb.
  - [x] No, I'm not able help debug or I don't have time to help.

**Comment by chuckremes** [comment]
I have access to an alpine docker image. Any chance you can provide a list of packages I need to `ask add` so I can build rubinius?

**Comment by chuckremes** [comment]
I added these pkgs:
`git, llvm, llvm-dev, clang, clang-dev, ruby`
The I installed bundler and ran `bundle install` in the cloned rubinius directory. It died trying to find `io/console`. Added `ruby-io-console` and bundle mostly worked. Died during melbourne build needing `make`. Now it needs `g++`. Finally, it died again because it can't find `ruby.h` to build melbourne.

I'm really stuck here.

**Comment by SquallATF** [comment]
@chuckremes `apk add clang llvm llvm-dev ncurses ncurses-dev bash openssl-dev sudo build-base linux-headers ruby ruby-dev ruby-rake ruby-bundler`
link sh to bash because testsuite not work will with ash. create a new user  set default shell with bash and allowed sudo , switch to new user and start build, configure is `./configure --prefix=/opt/rubinius/rubinius-3.70 --libc=libc.musl-x86_64.so.1`
some locale test will failed because musl lib c not support '[musl libc open issue](http://wiki.musl-libc.org/wiki/Open_Issues#C_locale_conformance)'

**Comment by chuckremes** [comment]
I have rubinius building and running on alpine. I get 42 spec failures upon build however. I get 11 spawn-related failures and 25 or so Time-related failures. Ugh... Time.

I'll try and figure out the spawn issues and fix them. Thanks for the report.

**Comment by chuckremes** [comment]
Oh boy, this was a doozy.

So the child process did record the proper errno (2 = ENOENT) but this errno got clobbered in the child prior to the child exiting. Therefore the parent picked up errno 22 (EINVAL) instead.

I traced this down to this line in rubinius:
https://github.com/rubinius/rubinius/blob/master/machine/logger.cpp#L243

There isn't anything special or weird about that line. However, the musl libc implementation of that basic function is generating an error somewhere internally and setting EINVAL. 

I have a fix which I will commit but it's contingent upon @brixen's approval. He might revert the fix.

My fix will be to save the errno in the child to another variable after execvp returns. The call to `exit` in the child will return this saved errno. In the parent I can retrieve the errno from the `status` var set by `waitpid`. I can then pass this errno to the appropriate logging functions, exceptions, etc.

The downside of this fix is that it will mask errors in the child process that occur after execvp returns. In this instance, it will mask an error in the logging mechanism.

**Comment by chuckremes** [comment]
BTW, if I can reproduce this reliably in a small C program with libmusl then I will open a ticket with them to get it fixed.

**Comment by chuckremes** [comment]
Fixed by d4c841e0bb.

**Comment by chuckremes** [comment]
Reported to musl's IRC channel. They do not have a bug tracker so they requested a report there. Here's repro code:
https://gist.github.com/chuckremes/698b0ee5d1279374cb1ef3add82a6f18

Also sent an email to their ML to report the issue. In the meantime, the fix above has resolved the issue.
